src/newton_solver.cpp

```diff
@@ -236,6 +262,43 @@ void NewtonSolverIterative::prepareLinearSystem(int ntry, int nnewt) {
                             "Linear solver SPBCG does not support sensitivity "
                             "computation for steady state problems.");
     }
+
+    // Get the Jacobian and its diagonal for preconditioning
+    model->fJ(*t, 0.0, *x, dx, xdot, ns_J.get());
+    model->fJDiag(*t, ns_Jdiag, 0.0, *x, dx);
+
+    // Ensure positivity of entries in ns_Jdiag
+    ns_p.set(1.0);
+    N_VAbs(ns_Jdiag.getNVector(), ns_Jdiag.getNVector());
+    N_VCompare(1e-15, ns_Jdiag.getNVector(), ns_tmp.getNVector());
+    N_VLinearSum(-1.0, ns_tmp.getNVector(), 1.0, ns_p.getNVector(), ns_tmp.getNVector());
+    N_VLinearSum(1.0, ns_Jdiag.getNVector(), 1.0, ns_tmp.getNVector(), ns_Jdiag.getNVector());
+}
+
+/* ------------------------------------------------------------------------- */
+
+void NewtonSolverIterative::prepareLinearSystemB(int ntry, int nnewt) {
+    newton_try = ntry;
+    i_newton = nnewt;
+    if (nnewt == -1) {
+        throw AmiException("Linear solver SPBCG does not support sensitivity "
+                           "computation for steady state problems.");
+    }
+
+    // Get the Jacobian and its diagonal for preconditioning
+    model->fJB(*t, 0.0, *x, dx, xB, dxB, xdot, ns_J.get());
+
+    // Get the diagonal and ensure negativity of entries is ns_J. Note that diag(JB) = -diag(J).
+    model->fJDiag(*t, ns_Jdiag, 0.0, *x, dx);
+
+    ns_p.set(1.0);
```

Comment:
duplicate code, move to seperate function?

Reply:
yikes. This code will be removed soon anyway, current issue, see #1073.
This is really an old handcrafted reimplementation of BICGSTAB. Prefer not to do changes now, but to fully remove and correctly reimplement the remaining linear solvers (i.e., all except DENSE and KLU).